llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp

```diff
@@ -1062,19 +1066,25 @@ bool MemCpyOptPass::performCallSlotOptzn(Instruction *cpyLoad,
   // If necessary, perform additional analysis.
   if (isModOrRefSet(MR))
     MR = BAA.callCapturesBefore(C, DestWithSrcSize, DT);
-  if (isModOrRefSet(MR))
+  if (isModOrRefSet(MR)) {
```

Comment:
Please add tests for these three cases which revert the code motion. I'm not sure an RAII guard is needed here. It will revert the change by default unless we explicitly keep the change in the end.

Reply:
RAII is a good idea. I have changed my implementation to an InstructionPositionGuard.